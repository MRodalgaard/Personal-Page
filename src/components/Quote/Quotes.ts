export interface IQuote {
  author: string;
  text: string;
}

export class Quotes {
  private quotes: IQuote[] = [
    {
      author: 'Steve Jobs',
      text:
        "You can't just ask customers what they want and then try to give that to them. By the time you get it built, they'll want something new.",
    },
    {
      author: 'Bill Gates',
      text: 'Your most unhappy customers are your greatest source of learning.',
    },
    {
      author: 'Albert Einstein',
      text: 'Life is like riding a bicycle. To keep your balance you must keep moving.',
    },
    {
      author: 'Thomas Edison',
      text:
        'Our greatest weakness lies in giving up. The most certain way to succeed is always to try just one more time.',
    },
    {
      author: 'Pablo Picasso',
      text: 'Every child is an artist. The problem is how to remain an artist once we grow up.',
    },
    {
      author: 'Pablo Picasso',
      text: 'Computers are useless. They can only give you answers.',
    },
    {
      author: 'Leonardo da Vinci',
      text: 'Simplicity is the ultimate sophistication.',
    },
    {
      author: 'Confucius',
      text:
        'Give a man a fish and he will eat for a day. Teach a man to fish and he will eat for a lifetime.',
    },
    {
      author: 'Dr. Seuss',
      text: "Don't cry because it's over, smile because it happened.",
    },
  ];

  public getRandomQuote = () => {
    const min = 0;
    const max = this.quotes.length - 1;
    const index = Math.floor(Math.random() * (max - min + 1)) + min;
    return this.quotes[index];
  };
}
