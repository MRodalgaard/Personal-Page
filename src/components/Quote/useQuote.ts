import { useState } from 'react';

export interface IQuote {
  author: string;
  text: string;
}

const quotes: IQuote[] = [
  {
    author: 'Steve Jobs',
    text:
      "You can't just ask customers what they want and then try to give that to them. By the time you get it built, they'll want something new.",
  },
  {
    author: 'Bill Gates',
    text: 'Your most unhappy customers are your greatest source of learning.',
  },
  {
    author: 'Albert Einstein',
    text:
      'Life is like riding a bicycle. To keep your balance you must keep moving.',
  },
  {
    author: 'Thomas Edison',
    text:
      'Our greatest weakness lies in giving up. The most certain way to succeed is always to try just one more time.',
  },
  {
    author: 'Pablo Picasso',
    text:
      'Every child is an artist. The problem is how to remain an artist once we grow up.',
  },
  {
    author: 'Pablo Picasso',
    text: 'Computers are useless. They can only give you answers.',
  },
  {
    author: 'Leonardo da Vinci',
    text: 'Simplicity is the ultimate sophistication.',
  },
  {
    author: 'Confucius',
    text:
      'Give a man a fish and he will eat for a day. Teach a man to fish and he will eat for a lifetime.',
  },
  {
    author: 'Dr. Seuss',
    text: "Don't cry because it's over, smile because it happened.",
  },
  {
    author: 'C.S. Lewis',
    text: 'Two of a trade never agree.',
  },
];

interface IProps {
  initialQuote?: IQuote;
}

const useQuote = ({ initialQuote }: IProps): [IQuote, () => void] => {
  const getRandomNumber = (max: number) => {
    return Math.floor(Math.random() * (max + 1));
  };

  const getNewRandomQuote = (previousQuote?: IQuote): IQuote => {
    const newQuote = quotes[getRandomNumber(quotes.length - 1)];
    if (newQuote.text === previousQuote?.text) {
      return getNewRandomQuote(previousQuote);
    }
    return newQuote;
  };

  const [quote, setQuote] = useState(initialQuote ?? getNewRandomQuote());

  const nextQuote = () => {
    setQuote(getNewRandomQuote(quote));
  };

  return [quote, nextQuote];
};

export default useQuote;
