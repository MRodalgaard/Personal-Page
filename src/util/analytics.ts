import 'firebase/analytics';
import firebase from './firebase';

let analytics: firebase.analytics.Analytics | undefined;

export enum LogEvent {
  ColorClick = 'color_click',
  QuoteClick = 'quote_click',
}

class Analytics {
  constructor() {
    firebase.analytics.isSupported().then((isSupported) => {
      if (isSupported) {
        analytics = firebase.analytics();
      }
    });
  }

  public logEvent = (event: LogEvent, parameters?: any) => {
    analytics?.logEvent(event, parameters);
  };
}

export default new Analytics();
