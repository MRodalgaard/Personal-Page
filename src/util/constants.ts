// Key for persisting app state in local storage
export const LOCALSTORAGE_KEY = "AppState";

export enum AppBackground {
  default,
  greyscale,
  colored
}
