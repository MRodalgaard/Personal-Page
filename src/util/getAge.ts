const getAge = (birthday: string): number | undefined => {
  const today = new Date();
  const birthDate = new Date(birthday);

  if (isNaN(birthDate.getTime()) || today < birthDate) {
    return undefined;
  }

  let age = today.getFullYear() - birthDate.getFullYear();
  const m = today.getMonth() - birthDate.getMonth();
  if (m < 0 || (m === 0 && today.getDate() < birthDate.getDate())) {
    age--;
  }
  return age;
};

export default getAge;
