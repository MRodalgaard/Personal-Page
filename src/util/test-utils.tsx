import { render, RenderOptions } from '@testing-library/react';
import AppProvider from 'components/App/AppProvider';
import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components';
import theme from './theme';

type Props = {
  children: ReactNode;
};

const AllTheProviders = ({ children }: Props) => {
  return (
    <AppProvider>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </AppProvider>
  );
};

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, {
    wrapper: AllTheProviders as React.FunctionComponent,
    ...options,
  });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
