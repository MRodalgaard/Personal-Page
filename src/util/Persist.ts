import { IAppState } from '../components/App/AppContext';
import { LOCALSTORAGE_KEY } from './constants';

const Persist = {
  persistState: (state: IAppState) => {
    localStorage.setItem(LOCALSTORAGE_KEY, JSON.stringify(state));
  },

  getPersistedState: (): IAppState | undefined => {
    const jsonState = localStorage.getItem(LOCALSTORAGE_KEY);
    return jsonState ? JSON.parse(jsonState) : undefined;
  },
};

export default Persist;
