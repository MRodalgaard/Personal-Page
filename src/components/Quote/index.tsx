import * as React from 'react';
import { useState } from 'react';
import styled from 'styled-components';
import analytics, { LogEvent } from '../../util/analytics';
import { AppColors } from '../../util/theme';
import { IQuote, Quotes } from './Quotes';

const Container = styled.div`
  border-left: 5px solid ${AppColors.grey};
  padding-left: 10px;
  user-select: none;
  transition: color 0.5s ease;

  &:hover {
    color: ${AppColors.secondary};
    cursor: pointer;
  }
`;

const AuthorText = styled.p`
  text-align: right;
  color: ${(props) => (props.color ? props.color : AppColors.primary)};
`;

interface IProps {
  color?: string;
  initialQuote?: IQuote;
}

const Quote = ({ color, initialQuote }: IProps) => {
  const quotes = new Quotes();
  const [quote, setQuote] = useState(
    initialQuote ? initialQuote : quotes.getRandomQuote()
  );

  const onClick = () => {
    setQuote(quotes.getRandomQuote());
    analytics.logEvent(LogEvent.QuoteClick);
  };

  return (
    <Container onClick={onClick}>
      <i>{`"${quote.text}"`}</i>
      <AuthorText color={color}>- {quote.author}</AuthorText>
    </Container>
  );
};

export default Quote;
