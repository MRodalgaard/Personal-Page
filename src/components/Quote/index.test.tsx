import React from 'react';
import { fireEvent, render, screen } from 'util/test-utils';
import Quote from '.';
import { IQuote } from './useQuote';

describe('Quote', () => {
  const quote: IQuote = {
    author: 'Author',
    text: 'Text',
  };
  const text = '"Text"';

  it('renders', () => {
    const { container } = render(<Quote initialQuote={quote} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('can select a different quote', () => {
    render(<Quote initialQuote={quote} />);
    expect(screen.queryByText(text)).toBeInTheDocument();
    fireEvent.click(screen.getByText(text));
    expect(screen.queryByText(text)).not.toBeInTheDocument();
  });
});
