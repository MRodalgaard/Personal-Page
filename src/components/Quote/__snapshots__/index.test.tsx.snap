// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quote renders 1`] = `
.c0 {
  border-left: 5px solid #333333;
  padding-left: 10px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: color 0.5s ease;
  transition: color 0.5s ease;
}

.c0:hover {
  color: #c9282d;
  cursor: pointer;
}

.c1 {
  text-align: right;
  color: #000000;
}

<div
  class="c0"
>
  <i>
    "Text"
  </i>
  <p
    class="c1"
  >
    - 
    Author
  </p>
</div>
`;
